java/ql/src/experimental/Security/CWE/CWE-352/JsonpInjectionLib.qll

```diff
@@ -0,0 +1,150 @@
+
+/**
+ * Taint-tracking configuration tracing flow from untrusted inputs to an argument of a function whose result is used as an if-statement condition.
+ *
+ * For example, in the context `String userControlled = request.getHeader("xyz"); boolean isGood = checkToken(userControlled); if(isGood) { ...`,
+ * the flow from `getHeader(...)` to the argument to `checkToken`, and then the flow from `checkToken`'s result to the condition of `if(isGood)`.
+ */
+class VerificationMethodFlowConfig extends TaintTracking2::Configuration {
+  VerificationMethodFlowConfig() { this = "VerificationMethodFlowConfig" }
+
+  override predicate isSource(DataFlow::Node src) { src instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(IfStmt is, Method m | is.getEnclosingCallable() = m |
+      (
+        not m.getAParameter().getName().regexpMatch("(?i).*(token|auth|referer|origin).*")
+        or
+        not m.getName().regexpMatch("(?i).*(token|auth|referer|origin).*")
+      ) and
+      sink.asExpr() = is.getCondition()
+    )
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node prod, DataFlow::Node succ) {
+    exists(MethodAccess ma |
+      (
+        ma.getMethod().getAParameter().getName().regexpMatch("(?i).*(token|auth|referer|origin).*")
+        or
+        ma.getMethod().getName().regexpMatch("(?i).*(token|auth|referer|origin).*")
+      ) and
+      ma.getAnArgument() = prod.asExpr() and
+      ma = succ.asExpr()
+    )
+  }
+}
+
+/** Get Callable by recursive method. */
+Callable getACallingCallableOrSelf(Callable call) {
+  result = call
+  or
+  result = getACallingCallableOrSelf(call.getAReference().getEnclosingCallable())
+}
+
+/**
+ * A method that is called to handle an HTTP GET request.
+ */
+abstract class RequestGetMethod extends Method { }
+
+/** Override method of `doGet` of `Servlet` subclass. */
+private class ServletGetMethod extends RequestGetMethod {
+  ServletGetMethod() { this instanceof DoGetServletMethod }
+}
+
+/** The method of SpringController class processing `get` request. */
+abstract class SpringControllerGetMethod extends RequestGetMethod { }
+
+/** Method using `GetMapping` annotation in SpringController class. */
+class SpringControllerGetMappingGetMethod extends SpringControllerGetMethod {
+  SpringControllerGetMappingGetMethod() {
+    this.getAnAnnotation()
+        .getType()
+        .hasQualifiedName("org.springframework.web.bind.annotation", "GetMapping")
+  }
+}
+
+/** The method that uses the `RequestMapping` annotation in the SpringController class and only handles the get request. */
+class SpringControllerRequestMappingGetMethod extends SpringControllerGetMethod {
+  SpringControllerRequestMappingGetMethod() {
+    this.getAnAnnotation()
+        .getType()
+        .hasQualifiedName("org.springframework.web.bind.annotation", "RequestMapping") and
+    this.getAnAnnotation().getValue("method").toString().regexpMatch("RequestMethod.GET|\\{...\\}") and
```

Comment:
Another use of `toString()` here -- this method should only be used to expose user-friendly representation. Will `.(CompileTimeConstantExpr).stringValue()` work here too?

Reply:
Received, let me take a look.

Reply:
@smowton I did a test, and the conclusion is that it cannot be used: `SpringControllerMethod.getAnAnnotation().getValue("method").(CompileTimeConstantExpr)` has no result.

Reply:
Oops, didn't realise it isn't string-typed. Try `.(VarAccess).getVariable().getName() = "GET"` instead.

It looks like you maybe also want to pick up RequestMappings without a `method` specified at all? For those you could use `.(ArrayInit).getSize() = 0` or `not getValue("method") instanceof VarAccess`.

Reply:
To find this sort of stuff out btw: try using `getAPrimaryQlClass` on an interesting expression (in this case `SpringControllerMethod.getAnAnnotation().getValue("method")`) to find out what classes it implements.

Reply:
get.